File: apps/api/tierzo_api/capabilities.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


Outcome = Literal["normal", "degraded"]
WarningCode = Literal[
    "openai_unconfigured_heuristic",
    "openai_provider_heuristic_fallback",
    "tmdb_unconfigured_text_fallback",
    "tmdb_provider_text_fallback",
    "tmdb_partial_match",
    "unsupported_planner_tool_text_fallback",
]
EffectiveMode = Literal[
    "deterministic",
    "openai",
    "heuristic",
    "tmdb",
    "unavailable",
]
ReasonCode = Literal["openai_unconfigured", "tmdb_unconfigured"]


WARNING_MESSAGES: dict[WarningCode, str] = {
    "openai_unconfigured_heuristic": (
        "OpenAI is not configured; Tierzo used deterministic planning."
    ),
    "openai_provider_heuristic_fallback": (
        "OpenAI planning was unavailable; Tierzo used deterministic planning."
    ),
    "tmdb_unconfigured_text_fallback": (
        "TMDb is not configured; Tierzo generated text cards instead."
    ),
    "tmdb_provider_text_fallback": (
        "TMDb lookup failed; Tierzo generated text cards instead."
    ),
    "tmdb_partial_match": (
        "Some movie posters were unavailable; Tierzo used text cards for them."
    ),
    "unsupported_planner_tool_text_fallback": (
        "The planned asset provider is not supported; Tierzo used text cards."
    ),
}


class ResultWarning(BaseModel):
    code: WarningCode
    message: str


class CapabilityEntry(BaseModel):
    available: bool
    effective_mode: EffectiveMode
    reason_code: ReasonCode | None = None


class Capabilities(BaseModel):
    text_cards: CapabilityEntry
    prompt_drafting: CapabilityEntry
    auto_planning: CapabilityEntry
    tmdb_movie: CapabilityEntry


class CapabilitiesResponse(BaseModel):
    schema_version: Literal["tierzo.capabilities.v1"] = "tierzo.capabilities.v1"
    capabilities: Capabilities


def make_warning(code: WarningCode) -> ResultWarning:
    return ResultWarning(code=code, message=WARNING_MESSAGES[code])


def deduplicate_warnings(
    warnings: list[ResultWarning],
) -> list[ResultWarning]:
    result: list[ResultWarning] = []
    seen: set[WarningCode] = set()
    for warning in warnings:
        if warning.code in seen:
            continue
        seen.add(warning.code)
        result.append(warning)
    return result


def outcome_for(warnings: list[ResultWarning]) -> Outcome:
    return "degraded" if warnings else "normal"


def build_capabilities(
    *,
    openai_configured: bool,
    tmdb_configured: bool,
) -> CapabilitiesResponse:
    planning = CapabilityEntry(
        available=True,
        effective_mode="openai" if openai_configured else "heuristic",
        reason_code=None if openai_configured else "openai_unconfigured",
    )
    return CapabilitiesResponse(
        capabilities=Capabilities(
            text_cards=CapabilityEntry(
                available=True,
                effective_mode="deterministic",
            ),
            prompt_drafting=planning,
            auto_planning=planning,
            tmdb_movie=CapabilityEntry(
                available=tmdb_configured,
                effective_mode="tmdb" if tmdb_configured else "unavailable",
                reason_code=None if tmdb_configured else "tmdb_unconfigured",
            ),
        )
    )

File: apps/web/lib/capabilities.ts
import type {
  CapabilitiesResponse,
  CapabilityEntry,
} from "./types.ts";

const MODES = new Set([
  "deterministic",
  "openai",
  "heuristic",
  "tmdb",
  "unavailable",
]);
const REASONS = new Set([
  "openai_unconfigured",
  "tmdb_unconfigured",
]);

export const unavailableCapabilities: CapabilitiesResponse = {
  schema_version: "tierzo.capabilities.v1",
  capabilities: {
    text_cards: {
      available: true,
      effective_mode: "deterministic",
      reason_code: null,
    },
    prompt_drafting: {
      available: true,
      effective_mode: "heuristic",
      reason_code: "openai_unconfigured",
    },
    auto_planning: {
      available: true,
      effective_mode: "heuristic",
      reason_code: "openai_unconfigured",
    },
    tmdb_movie: {
      available: false,
      effective_mode: "unavailable",
      reason_code: "tmdb_unconfigured",
    },
  },
};

export function parseCapabilities(value: unknown): CapabilitiesResponse {
  if (!isRecord(value) || value.schema_version !== "tierzo.capabilities.v1") {
    throw new Error("Tierzo received an invalid capabilities response.");
  }
  const capabilities = value.capabilities;
  if (
    !isRecord(capabilities) ||
    !isCapabilityEntry(capabilities.text_cards) ||
    capabilities.text_cards.effective_mode !== "deterministic" ||
    !capabilities.text_cards.available ||
    capabilities.text_cards.reason_code !== null ||
    !isCapabilityEntry(capabilities.prompt_drafting) ||
    !["openai", "heuristic"].includes(
      capabilities.prompt_drafting.effective_mode,
    ) ||
    !capabilities.prompt_drafting.available ||
    !hasPlanningReason(capabilities.prompt_drafting) ||
    !isCapabilityEntry(capabilities.auto_planning) ||
    !["openai", "heuristic"].includes(
      capabilities.auto_planning.effective_mode,
    ) ||
    !capabilities.auto_planning.available ||
    !hasPlanningReason(capabilities.auto_planning) ||
    !isCapabilityEntry(capabilities.tmdb_movie) ||
    !["tmdb", "unavailable"].includes(
      capabilities.tmdb_movie.effective_mode,
    ) ||
    capabilities.tmdb_movie.available !==
      (capabilities.tmdb_movie.effective_mode === "tmdb") ||
    (capabilities.tmdb_movie.available
      ? capabilities.tmdb_movie.reason_code !== null
      : capabilities.tmdb_movie.reason_code !== "tmdb_unconfigured")
  ) {
    throw new Error("Tierzo received an invalid capabilities response.");
  }
  return value as CapabilitiesResponse;
}

function hasPlanningReason(entry: CapabilityEntry): boolean {
  return entry.effective_mode === "openai"
    ? entry.reason_code === null
    : entry.reason_code === "openai_unconfigured";
}

function isCapabilityEntry(value: unknown): value is CapabilityEntry {
  if (!isRecord(value)) {
    return false;
  }
  return (
    typeof value.available === "boolean" &&
    typeof value.effective_mode === "string" &&
    MODES.has(value.effective_mode) &&
    (value.reason_code === null ||
      (typeof value.reason_code === "string" &&
        REASONS.has(value.reason_code)))
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
